### PR TITLE
Sponskrub integration

### LINKED ISSUE #209: [Site Request] Downloading a YouTube video by cutting out sponsors
## Checklist

- [x] I'm reporting a site feature request
- [x] I've verified that I'm running youtube-dlc version **2020.10.31**
- [x] I've searched the bugtracker for similar site feature requests including closed ones

## Description

Add an option to download a YouTube video by cutting out sponsered segments which are available in the [SponsorBlock API](https://github.com/ajayyy/SponsorBlock).

### PR DESCRIPTION
**Pull request by pukkandan**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Use [SponSkrub](https://github.com/faissaloo/SponSkrub) to cut out sponsor segments from youtube videos. Closes #209 

~~This is currently just a proof of concept. It needs more testing and refinement before being ready for use.~~

    --sponskrub                      Use sponskrub to mark sponsored sections
                                     with the data available in SponsorBlock API
                                     (Youtube only)
    --sponskrub-cut                  Cut out the sponsor sections instead of
                                     simply marking them
    --sponskrub-force                Run sponskrub even if the video was
                                     already downloaded. Use with caution
    --sponskrub-location             Location of the sponskrub binary;
                                     either the path to the binary or its
                                     containing directory
    --sponskrub-args                 Give these arguments to sponskrub

When using `--sponskrub-cut`, the postprocessor is run only if the video was downloaded/merged in the current run. This ensures that the the video does not get cut at sponsor timestamps multiple times. 

If the user (for whatever reason) wants to run the postprocessor on existing files, `--sponskrub-force` can be used.

### CODE REVIEW

**Comment by shakeyourbunny** [comment]
Note that there is no binary for windows there.

**Comment by pukkandan** [comment]
I need some help with a few things:
1. What should be the default. Currently it is off by default, but it can be set to on if the executable exists.
2. What should be the default arguments? Maybe `-chapter`? That will ensure it is fast and non-destructive.
3. SponsKrub is pointless for non-youtube videos. But I can't seem to find `ie_key` in the `information` object. Is there another way to quickly and  reliably tell if it is a youtube video?

4. If the download is run multiple times, the post-processors are re-run each time. When sponskrub is used without `-chapter`, this causes the video to get cut up again and again at the sponsor timestamps. This is a major issue. I have no idea how to fix this. Anybody has any ideas?

**Comment by pukkandan** [comment]
Also, I am testing the code only in x64 windows Cpython 3.8.4
I'd appreciate it if someone can test it in other systems.

File: youtube_dlc/postprocessor/sponskrub.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        if p.returncode == 0:
+            os.remove(filename)
+            os.rename(temp_filename, filename)
+            self._downloader.to_screen('[sponskrub] Sponsor sections have been marked/removed')
+        elif p.returncode != 3:  # error code 3 means there was no info about the video
+            stderr = stderr.decode('utf-8', 'replace')
+            msg = stderr.strip().split('\n')[-1]
+            raise PostProcessingError(msg if msg else 'Sponskrub failed!')
+        else:
+            self._downloader.to_screen('[sponskrub] No data in database')
```
Comment (ajayyy) [inline youtube_dlc/postprocessor/sponskrub.py:74]:
```suggestion
            self._downloader.to_screen('[sponskrub] No segments in the SponsorBlock database')
```

File: youtube_dlc/options.py [hunk 2]
```diff
@@ -878,6 +878,23 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
         metavar='FORMAT', dest='convertsubtitles', default=None,
         help='Convert the subtitles to other format (currently supported: srt|ass|vtt|lrc)')
 
+    extractor = optparse.OptionGroup(parser, 'SponSkrub Options')
```
Comment (ajayyy) [inline youtube_dlc/options.py:881]:
```suggestion
    extractor = optparse.OptionGroup(parser, 'SponSkrub Options (SponsorBlock)')
```

**Review by ajayyy — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
> If the download is run multiple times, the post-processors are re-run each time. When sponskrub is used without -chapter, this causes the video to get cut up again and again at the sponsor timestamps. This is a major issue. I have no idea how to fix this. Anybody has any ideas?

I can't complete the feature without some help in this matter.

**Comment by JanPokorny** [comment]
@pukkandan One possible solution: modify SponSkrub (add a config flag) so it checks the video length and no-ops if it doesn't match. Would that work?

**Comment by pukkandan** [comment]
@JanPokorny It is a good idea. However, I don't know any DLang (in which SponsKrub is written). Let me see if this can be implemented in the PostProcessor itself.

**Comment by JanPokorny** [comment]
Maybe @faissaloo wants to implement this feature for SponSkrub? IDK if the SponsorBlock API returns the video length, but if it does, that should be an easy enough change. I could see this being useful outside of this issue, since checking the video length before cutting it up seems like a reasonably general sanity check.

**Comment by faissaloo** [comment]
>     4\. If the download is run multiple times, the post-processors are re-run each time. When sponskrub is used without `-chapter`, this causes the video to get cut up again and again at the sponsor timestamps. This is a major issue. I have no idea how to fix this. Anybody has any ideas?

Trying to understand the context here, what would be a likely reason for running the download multiple times? I assume a download is done to completion right?

**Comment by pukkandan** [comment]
@faissaloo If a playlist download is run multiple times without using archive, for example.
youtube-dlc will see that the video file already exists and won't download it again. But the postprocessors WILL be run again

**Comment by JanPokorny** [comment]
@pukkandan That seems like a weird behavior tbh, I would expect that filters aren't supposed to be idempotent. Maybe @blackjack4494 can tell us more on how it works?

**Comment by faissaloo** [comment]
@pukkandan 
>  youtube-dlc will see that the video file already exists and won't download it again. But the postprocessors WILL be run again

I have to agree with @JanPokorny, this seems like an issue with youtube-dlc. I wouldn't expect such a thing to be in the purview of the application doing the processing, rather the application managing the download should determine for itself whether or not it has downloaded something.

**Comment by pukkandan** [comment]
@JanPokorny @faissaloo I have implemented a solution for this by ensuring that the postprocessor is run only if the video was downloaded/merged in the current run. A solution based on file duration is superior imo, but I'll leave that for someone else to implement.

I am also turning it on by default if sponskrub binary can be found. That is how mergerPP works, so it makes sense to do the same here.

**Comment by shivasagarrao** [comment]
@pukkandan Small request is try to implement this feature in [youtube-dl](https://github.com/ytdl-org/youtube-dl)  as well.

**Comment by pukkandan** [comment]
@shivasagarrao This patch can be used in youtube-dl without any changes. But I honestly am not interested in making any PRs over there. Feel free to cherry-pick my commits and make a PR over there yourself

**Pull request closed without merging**